### PR TITLE
created structure for whitelisting directories for govet shadow testing

### PR DESCRIPTION
**Pull request by stevekuznetsov**
@liggitt PTAL

follow-up pulls expand the whitelist `grep` and fix shadowed variables

### CODE REVIEW

File: hack/verify-govet.sh [hunk 1]
```diff
@@ -5,7 +5,7 @@ set -o pipefail
 
 GO_VERSION=($(go version))
 
-if [[ -z $(echo "${GO_VERSION[2]}" | grep -E 'go1.4') ]]; then
+if [[ -z $(echo "${GO_VERSION[2]}" | grep -E 'go1.5') ]]; then
```
Comment (stevekuznetsov) [inline hack/verify-govet.sh]: flags have changed for `go tool vet` and I think we can migrate testing to `1.5.*`

Comment (liggitt) [inline hack/verify-govet.sh]: Jenkins doesn't run on 1.5 yet, so if you want this actually enforced, it has to be 1.4 as well

File: hack/verify-govet.sh [hunk 2]
```diff
@@ -23,7 +23,55 @@ FAILURE=false
...
+./pkg/router
+./pkg/sdn
+./pkg/security
+./pkg/service
+./pkg/serviceaccounts
+./pkg/template
+./pkg/user
+./pkg/util
+./pkg/version
+./plugins
```
Comment (liggitt) [inline hack/verify-govet.sh]: I'd rather blacklist so we can delete exceptions as we clean them up and new packages will get enforced.

**Comment by stevekuznetsov** [comment]
[test]

File: hack/verify-govet.sh (hunk 2 above)
Comment (stevekuznetsov) [inline hack/verify-govet.sh]: :+1:

**Comment by stevekuznetsov** [comment]
@liggitt made changes to programatically generate `ALL_DIRS`

File: hack/verify-govet.sh [hunk 3]
```diff
@@ -23,13 +23,69 @@ FAILURE=false
 test_dirs=$(find_files | cut -d '/' -f 1-2 | sort -u)
 for test_dir in $test_dirs
 do
-  go tool vet -shadow=false $test_dir
+  go tool vet -all $test_dir
```
Comment (liggitt) [inline hack/verify-govet.sh]: should still be -shadow=false, right?

File: hack/verify-govet.sh [hunk 4]
```diff
@@ -23,13 +23,69 @@ FAILURE=false
...
+./pkg/util
+./pkg/version
+./plugins
+./test'
+
+for test_dir in $ALL_DIRS
+do
+  # use `grep` failure to determine that a directory is not in the blacklist 
+  if ! echo "${DIR_BLACKLIST}" | grep -q "${test_dir}"; then
+    go tool vet -shadow -shadowstrict $test_dir
```
Comment (liggitt) [inline hack/verify-govet.sh]: does `vet` recurse subpackages? If not, then you need ALL_DIRS to be all packages, and DIR_BLACKLIST to filter out by prefix

File: hack/verify-govet.sh (hunk 3 above)
Comment (stevekuznetsov) [inline hack/verify-govet.sh]: no. the flags to `go vet` changed with the latest release.
`-all` == all, *except* `shadow`

File: hack/verify-govet.sh (hunk 4 above)
Comment (stevekuznetsov) [inline hack/verify-govet.sh]: `vet` recurses directories, which is what we are passing it

**Comment by stevekuznetsov** [comment]
flake on `e2e` test [here](https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin/8143/consoleFull):
```
Error from server: 501: All the given peers are not reachable (failed to propose on members [https://172.18.11.218:4001] twice [last error: Unexpected HTTP status code]) [0]
```

#6447 

re[test]

**Comment by stevekuznetsov** [comment]
Another etcd flake [here](https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin/8146/console):
```
Error from server: deploymentConfig "failing-dc" cannot be updated: the object has been modified; please apply your changes to the latest version and try again
```

#6447 
re[test]

**Comment by stevekuznetsov** [comment]
end-to-end flake [here](https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin/8150/console):
```
Error from server: 501: All the given peers are not reachable (failed to propose on members [https://172.18.7.126:4001] twice [last error: Unexpected HTTP status code]) [0]
```

#6447 
re[test]

**Comment by stevekuznetsov** [comment]
flake on `TestAuthorizationRestrictedAccessForProjectAdmins` [here]():
```
F0104 16:11:17.513223   22923 run_components.go:393] Unable to initialize namespaces: unable to persist the updated namespace UID allocations: uidallocation "" cannot be updated: another caller has already initialized the resource
```

#6065 
re[test]

**Comment by stevekuznetsov** [comment]
more flakes with webdriver [here](https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin/8167/console):
```
1)  unauthenticated user should be able to log in
  - Error: Angular could not be found on the page https://localhost:9000/ : angular never provided resumeBootstrap
...
2)  authenticated e2e-user with test project should be able to list the test project
  - Error: Error while waiting for Protractor to sync with the page: "angular could not be found on the window"
...
!!! Error in hack/../test/end-to-end/core.sh:346
	'grunt test-integration' exited with status 3
Call stack:
	1: hack/../test/end-to-end/core.sh:346 main(...)
Exiting with status 1
```
https://github.com/openshift/origin/issues/6533
re[test]

File: hack/verify-govet.sh [hunk 5]
```diff
@@ -30,6 +30,62 @@ do
...
+    if [ "$?" -ne 0 ]
```
Comment (liggitt) [inline hack/verify-govet.sh]: seems odd to compare a string and a number... this probably works, but make both sides one or the other

**Comment by liggitt** [comment]
nit, LGTM

**Comment by stevekuznetsov** [comment]
@liggitt addressed comment

**Comment by stevekuznetsov** [comment]
@liggitt this is ready

**Comment by liggitt** [comment]
[merge]

**Pull request merged**